### ISSUE TITLE
truth value of a series is ambiguous for boolean comparison 

### ISSUE THREAD

**Issue by johnjhawk** [InvestmentSystems/static-frame#236]
Hi, 
Inadvertently using a boolean operator to compare series is easy to do and there is currently no warning or failure in static frame. In the below example, static frame is doing the correct thing, but without failing/warning, it's easy for the user it get unintended results. 
For example,
```
<FrameGO>
<IndexGO> a       b       c       <<U1>
<Index>
0         1       1       1
1         1       1       2
2         1       1       1
<int64>   <int64> <int64> <int64>
```
The result below is expected but can be confusing as the logical operator is evaluating if the arrays are empty or not and then returning the second result.  I expect all rows to be false in this case. 
```
In [51]: (f1['a'] != f1['b']) and (f1['a'] == f1['c'])                                                                                                                                        
Out[51]: 
<Series>
<Index>
0        True
1        False
2        True
<int64>  <bool>
```
Using the bitwise operator is the correct way to compare these series at the element level, but it's easy to use 'and' instead of "&' and get a result that looks okay but is different than expected. 
```
In [52]: (f1['a'] != f1['b']) & (f1['a'] == f1['c'])                                                                                                                                          
Out[52]: 
<Series>
<Index>
0        False
1        False
2        False
<int64>  <bool>
```
Have you thought about raising or warning when this potentially ambiguous situation arises? Note that numpy raises a ValueError in this situation.

Thanks,
John

**Comment by ForeverWintr** [comment]
This issue rears its head in other places too, for example:

```python
assert sf.Series([1,2,3]) == sf.Series([1,2,4])
```

Looks like it's checking that the two series are equal, but will actually always be true (as will `assert sf.Series([1,2,3]) == 5`, etc.)

I think we should follow `numpy` and `pandas`' lead here and raise in these situations:

```python
assert np.array([1, 2, 3]) == [1, 2, 4]
Traceback (most recent call last):
  Python Shell, prompt 12, line 1
builtins.ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

**Comment by chaburkland** [comment]
This is a good find. I think it's because truthiness is exclusively defined as whether or not the frame/series have any values (source code is `return bool(self.values.size)`). This means that Python `and` will have unexpected behavior, since any non-empty Frames/Series will always be True.

In John's example, both sides of the `and` operator are truthy, meaning it's going to return the second object, hence the `[True, False, True]` series.

In Tom's example, since the Series is non-empty, it will always pass assert checks.

I agree with Tom - this needs to raise :+1:

**Comment by flexatone** [comment]
Many thanks for initiating this discussion.

I have been ambivalent about this issue in the past, but these examples highlight that a more defensive approach is sensible. As reducing opportunities for error is a goal of StaticFrame, I support this change.

As this is a change to the API, we can incorporate it in the 0.7 release.

**Comment by flexatone** [comment]
This has been implemented on the api/070 branch, and will be merged for the release of 0.7.0.
